azext_iot/central/providers/device_provider.py

```diff
@@ -0,0 +1,79 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from azext_iot.central import services as central_services
+
+
+class CentralDeviceProvider:
+    def __init__(self, cmd, app_id, token=None):
+        """
+        Provider for device/device_template APIs
+
+        Args:
+            cmd: command passed into az
+            app_id: name of app (used for forming request URL)
+            token: (OPTIONAL) authorization token to fetch device details from IoTC.
+                MUST INCLUDE type (e.g. 'SharedAccessToken ...', 'Bearer ...')
+                Useful in scenarios where user doesn't own the app
+                therefore AAD token won't work, but a SAS token generated by owner will
+        """
+        self._cmd = cmd
+        self._app_id = app_id
+        self._token = token
+        self._device_templates = {}
+        self._devices = {}
+
+    def get_device_template(
+        self, device_id, central_dns_suffix="azureiotcentral.com",
+    ):
+        device = self.get_device(device_id, central_dns_suffix)
+        device_template_urn = device["instanceOf"]
+
+        if not device_template_urn:
+            raise ValueError(
+                "No device template urn found for device '{}'".format(device_id)
+            )
+
+        # get or add to cache
+        if (
+            device_template_urn not in self._device_templates
+            or not self._device_templates.get(device_template_urn)
+        ):
+            self._device_templates[
+                device_template_urn
+            ] = central_services.device_template.get_device_template(
+                self._cmd,
+                device_template_urn,
+                self._app_id,
+                self._token,
+                central_dns_suffix,
+            )
+
+        device_template = self._device_templates.get(device_template_urn)
```

Comment:
`.get()` is slower than `self._device_templates[device_template_urn]`.  You know the key exists in the dictionary from `get_device_template` right?

Reply:
oh good catch, I think if the code reaches here and its still missing from catch then we should throw a KeyError and let the handler handle it